workflows/namespace.go

```diff
@@ -0,0 +1,307 @@
+	)
+	defer func() {
+		if err != nil {
+			out.setError(err)
+		}
+		if ns != nil {
+			out.Namespace = ns
+		} else if spec != nil {
+			out.Namespace = &namespace.Namespace{
+				Namespace: namespaceID,
+				Spec:      spec,
+			}
+		}
+	}()
+	if ns == nil {
+		var createResp *cloudservice.CreateNamespaceResponse
+		// no namespace found, create one
+		createResp, err = w.CreateNamespace(ctx, &cloudservice.CreateNamespaceRequest{
+			Spec: spec,
+		})
+		if err != nil {
+			return out, err
+		}
+		namespaceID = createResp.Namespace
+		asyncOpID = createResp.AsyncOperation.Id
+		out.Outcome = ReconcileOutcomeCreated
+
+	} else if !proto.Equal(ns.Spec, spec) {
+		var updateResp *cloudservice.UpdateNamespaceResponse
+		// namespace found, and specs don't match,  update it
+		updateResp, err = w.UpdateNamespace(ctx, &cloudservice.UpdateNamespaceRequest{
+			Namespace:       ns.Namespace,
+			Spec:            spec,
+			ResourceVersion: ns.ResourceVersion,
+		})
+		if err != nil {
+			return out, err
+		}
+		namespaceID = ns.Namespace
+		asyncOpID = updateResp.AsyncOperation.Id
+		out.Outcome = ReconcileOutcomeUpdated
+
+	} else {
+		// nothing to change, get the latest namespace and return
+		namespaceID = ns.Namespace
+		out.Outcome = ReconcileOutcomeUnchanged
+		return out, nil
+	}
+
+	if asyncOpID != "" {
+		// wait for the operation to complete
+		_, err = w.WaitForAsyncOperation(ctx, &WaitForAsyncOperationInput{
+			AsyncOperationID: asyncOpID,
+			Timeout:          namespaceUpdateTimeout,
+		})
+		if err != nil {
+			return out, err
+		}
+	}
+	var getResp *cloudservice.GetNamespaceResponse
+	getResp, err = w.GetNamespace(ctx, &cloudservice.GetNamespaceRequest{
+		Namespace: namespaceID,
+	})
+	if err != nil {
+		return out, err
+	}
+	ns = getResp.Namespace
+	return out, nil
+}
+
+// Reconcile a namespace, create the namespace if one does not exist, or update the namespace if one does exist.
+func (w *workflows) ReconcileNamespace(ctx workflow.Context, in *ReconcileNamespaceInput) (*ReconcileNamespaceOutput, error) {
+	if err := validator.ValidateStruct(in); err != nil {
+		return nil, fmt.Errorf("invalid input: %s", err)
+	}
+	namespace, err := w.GetNamespaceWithName(ctx, in.Spec.Name)
+	if err != nil {
+		return nil, err
+	}
+	out, err := w.reconcileNamespace(ctx, in.Spec, namespace)
+	if err != nil {
+		return nil, err
+	}
+	return out, nil
+}
+
+// Reconcile a namespace, create the namespace if one does not exist, or update the namespace if one does exist.
+func (w *workflows) ReconcileNamespaces(ctx workflow.Context, in *ReconcileNamespacesInput) (*ReconcileNamespacesOutput, error) {
+	if err := validator.ValidateStruct(in); err != nil {
+		return nil, fmt.Errorf("invalid input: %s", err)
+	}
+	var (
+		getNamespacesReq = &cloudservice.GetNamespacesRequest{}
+		namespaces       = make(map[string]*namespace.Namespace)
+		out              = &ReconcileNamespacesOutput{}
+	)
+	for {
+		resp, err := w.GetNamespaces(ctx, getNamespacesReq)
```

Comment:
doesn't GetNamespaces already do the pagination though? why do we have to do it again?

Reply:
yes you are right, fixed.